File: backend/app/schemas/maschine_auslastung.py
"""Schemas für Maschinenauslastung."""

from __future__ import annotations

from pydantic import BaseModel, Field


class MaschineAuslastungProjectContribution(BaseModel):
    project_id: int
    project_name: str
    source_type: str
    source_label: str
    jahresstueckzahl: float
    required_hours: float


class MaschineAuslastungYearRow(BaseModel):
    calendar_year: int
    required_hours: float
    available_hours: float | None
    utilization_pct: float | None


class MaschineAuslastungRow(BaseModel):
    maschine_id: int
    maschinen_nr: str
    bezeichnung: str
    werk_id: int | None
    werk_name: str | None
    available_hours: float | None
    required_hours: float
    utilization_pct: float | None
    rest_capacity_hours: float | None
    overload_hours: float | None
    is_overloaded: bool
    has_demand: bool
    projects: list[MaschineAuslastungProjectContribution] = Field(default_factory=list)
    yearly_breakdown: list[MaschineAuslastungYearRow] = Field(default_factory=list)


class MaschineAuslastungPlanningPeriod(BaseModel):
    label: str
    basis: str
    available_hours_per_machine_year: float | None


class MaschineAuslastungSummary(BaseModel):
    machine_count: int
    machines_with_demand: int
    overloaded_count: int
    average_utilization_pct: float | None
    plant_utilization_pct: float | None
    max_utilization_pct: float | None
    max_utilization_maschine_id: int | None
    max_utilization_maschine_name: str | None


class MaschineAuslastungResponse(BaseModel):
    plant_id: int
    plant_name: str
    customer_id: int | None
    program_id: int | None
    project_ids: list[int]
    no_projects_selected: bool
    planning_period: MaschineAuslastungPlanningPeriod
    summary: MaschineAuslastungSummary
    machines: list[MaschineAuslastungRow]

File: frontend/src/types/maschineAuslastung.ts
export interface MaschineAuslastungProjectContribution {
  project_id: number;
  project_name: string;
  source_type: string;
  source_label: string;
  jahresstueckzahl: number;
  required_hours: number;
}

export interface MaschineAuslastungYearRow {
  calendar_year: number;
  required_hours: number;
  available_hours: number | null;
  utilization_pct: number | null;
}

export interface MaschineAuslastungRow {
  maschine_id: number;
  maschinen_nr: string;
  bezeichnung: string;
  werk_id: number | null;
  werk_name: string | null;
  available_hours: number | null;
  required_hours: number;
  utilization_pct: number | null;
  rest_capacity_hours: number | null;
  overload_hours: number | null;
  is_overloaded: boolean;
  has_demand: boolean;
  projects: MaschineAuslastungProjectContribution[];
  yearly_breakdown: MaschineAuslastungYearRow[];
}

export interface MaschineAuslastungPlanningPeriod {
  label: string;
  basis: string;
  available_hours_per_machine_year: number | null;
}

export interface MaschineAuslastungSummary {
  machine_count: number;
  machines_with_demand: number;
  overloaded_count: number;
  average_utilization_pct: number | null;
  plant_utilization_pct: number | null;
  max_utilization_pct: number | null;
  max_utilization_maschine_id: number | null;
  max_utilization_maschine_name: string | null;
}

export interface MaschineAuslastungResponse {
  plant_id: number;
  plant_name: string;
  customer_id: number | null;
  program_id: number | null;
  project_ids: number[];
  no_projects_selected: boolean;
  planning_period: MaschineAuslastungPlanningPeriod;
  summary: MaschineAuslastungSummary;
  machines: MaschineAuslastungRow[];
}

export interface MaschineAuslastungParams {
  plant_id: number;
  customer_id?: number;
  program_id?: number;
  project_ids?: number[];
}
